### PR TITLE
Implement Issue Config

### LINKED ISSUE #20955: Disable blank Issues
### Feature Description

Gitea allows providing Issue templates. If a Repo has Issue templates, the User can choose between the Templates and a blank Issue, there is currently no way to disable the blank Issue. GitHub allows this and there are a lot of Repos out there who are using this feature.

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by JakobDev**
Closes #20955

This PR adds the possibility to disable blank Issues, when the Repo has templates. This can be done by creating the file `.gitea/issue_config.yaml` with the content `blank_issues_enabled` in the Repo.

I have no experience in Go, so this code might not be perfect, but it is working. If anybody knows how to do that better, just say it. It is also missing documentation at the moment.

### CODE REVIEW

**Comment by lunny** [comment]
should the path be `.gitea/issue_template/config.yaml` and to keep compatible with Github to support `.github/issue_template/config.yaml`

**Comment by JakobDev** [comment]
@lunny I have now changed the paths

File: modules/context/repo.go
Comment (lunny) [inline modules/context/repo.go:1170]: Why not return err?

Comment (JakobDev) [reply modules/context/repo.go:1170]: I think, if anything goes wrong here, three is nothing a normal User can do. So we should not show it in the GUI.

Comment (lunny) [reply modules/context/repo.go:1170]: Imo, the error should be posted to UI since we cannot read default branch.

**Comment by JakobDev** [comment]
@lunny Any more suggestions or should I add documentation, so this can be merged?

File: modules/context/repo.go
Comment (delvh) [inline modules/context/repo.go]: Perhaps we should rewrite that to **one** regex?
Otherwise, this will scale **really** poorly.

**Comment by lunny** [comment]
Hi, sorry to be late to reply. Maybe you could add some documentation and tests? And maybe @wolfogre could also have a review because he contributed some related PRs.

**Comment by JakobDev** [comment]
The documentation has now been added, but i don't know, how to write tests for this.

File: routers/web/repo/issue.go
Comment (delvh) [inline routers/web/repo/issue.go]: Why open a new key, when we can also reuse `ctx.Flash.Error()`?

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (delvh) [inline docs/content/doc/usage/issue-pull-request-templates.en-us.md]:
```suggestion
| BlankIssuesEnabled | If set to false, the User is forced to use a Template                                                 | Boolean |
```

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (delvh) [inline docs/content/doc/usage/issue-pull-request-templates.en-us.md]: I think `Default` is also a good idea for this table.

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (delvh) [reply docs/content/doc/usage/issue-pull-request-templates.en-us.md]: Also, on GitHub, this is written in snake case: `blank_issues_enabled: false`.
Does our YAML parser parse all text cases, or do we need to adapt something?
And if it allows multiple text cases, we should probably document that.

File: templates/repo/issue/choose.tmpl
Comment (delvh) [inline templates/repo/issue/choose.tmpl]: This won't be needed anymore with `ctx.Flash`.

File: routers/web/repo/issue.go
Comment (delvh) [reply routers/web/repo/issue.go]: Or is that not available for issues?
If that's the case, we should probably fix it.

File: modules/context/repo.go
Comment (wolfogre) [inline modules/context/repo.go]: Should use `yaml.v3` instead of `yaml.v2`, see #21832

File: modules/context/repo.go
Comment (wolfogre) [reply modules/context/repo.go]: I don't think so, they are not patterns but paths, to check if the file exists one by one.

It's inefficient to traverse all files and find a match for the pattern.

File: modules/context/repo.go
Comment (wolfogre) [inline modules/context/repo.go]: The check isn't necessary, `template.CouldBe` will return false if it's config file.

File: modules/context/repo.go
Comment (wolfogre) [inline modules/context/repo.go]: Why not `(*api.IssueConfig, error)` and return `nil, err` if there's a error?

File: modules/context/repo.go
Comment (wolfogre) [inline modules/context/repo.go:1103]: Why is the default true?

File: modules/context/repo.go
Comment (wolfogre) [reply modules/context/repo.go]: And I don't think we need to define this function, the body is too simple.

See:

```go
		issueConfig, err := ExtractIssueConfigFromYaml(configContent)
		if err != nil {
			return defaultIssueConfig, err
		}
```

```go
		issueConfig := api.IssueConfig{}
		if err := yaml.Unmarshal(configContent, &issueConfig); err != nil {
			return defaultIssueConfig, err
		}
```

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (wolfogre) [reply docs/content/doc/usage/issue-pull-request-templates.en-us.md]: It should be `blank_issues_enabled`.

The [definition](https://github.com/go-gitea/gitea/pull/20956/files#diff-828ef82cb24a01a8bb0409e2bbb87296e058c1c3e78d45d599e800268ee7d2b8R194) named it with tag `yaml:"blank_issues_enabled"`.

File: routers/web/repo/issue.go
Comment (JakobDev) [reply routers/web/repo/issue.go]: `ctx.Flash.Error("Test123")` does nothing, so I'm guess it is not available here

Comment (JakobDev) [reply routers/web/repo/issue.go]: Now I see it. It's on the main repo page not the Issues one. Is that really the right place for this?

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (JakobDev) [reply docs/content/doc/usage/issue-pull-request-templates.en-us.md]: It is `blank_issues_enabled`. This is a mistake of me while writing the documentation. I will fix it in the next commit.

File: modules/context/repo.go
Comment (JakobDev) [reply modules/context/repo.go:1103]: The blank Issues should be enabled, if nothing other specified

File: modules/context/repo.go
Comment (JakobDev) [reply modules/context/repo.go]: > Why not (*api.IssueConfig, error) and return nil, err if there's a error?

Return the default config makes life a little bit easier

File: routers/web/repo/issue.go
Comment (JakobDev) [reply routers/web/repo/issue.go]: From testing, ctx.Flash.Error causes a few problems. I stay with the current solution.

Comment (delvh) [reply routers/web/repo/issue.go]: Then please add a comment to both the template and the line `	ctx.Data["IssueConfigError"] = err`, then I'm fine with it.

**Comment by JakobDev** [comment]
I had now done a bit of code refactoring:
- I now use yaml.v3
- The list with the filenames is shorten
- If a Issue config contains a error, it is now shown while viewing the file

`ctx.Flash.Error()` requires loading the config every time you visit the main page of the repo (which is a waste of resources), or else it will just be shown after visiting, the Issue template list, which is confusing, so I decided to stick with the current solution.

File: routers/web/repo/issue.go
Comment (delvh) [reply routers/web/repo/issue.go]: Something like
```suggestion
	ctx.Data["IssueConfigError"] = err // ctx.Flash.Err makes problems here
```

File: templates/repo/issue/choose.tmpl
Comment (delvh) [inline templates/repo/issue/choose.tmpl]:
```suggestion
		{{- if .IssueConfigError}}{{/* normal warning flash makes problems here*/}}
```

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (delvh) [inline docs/content/doc/usage/issue-pull-request-templates.en-us.md]:
```suggestion
blank_issues_enabled: true
```

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (delvh) [inline docs/content/doc/usage/issue-pull-request-templates.en-us.md]:
```suggestion
| Key                  | Description                                                                                           | Type    | Default |
|----------------------|-------------------------------------------------------------------------------------------------------|---------|---------|
| blank_issues_enabled | If set to false, the User is forced to use a Template                                                 | Boolean | true |
```

File: modules/context/repo.go
Comment (delvh) [inline modules/context/repo.go:1110]:
```suggestion
// GetIssueConfig returns the issue config for this repo.
// It never returns a nil config.
func (r *Repository) GetIssueConfig(path string, commit *git.Commit) (api.IssueConfig, error) {
```

File: templates/repo/issue/choose.tmpl
Comment (delvh) [inline templates/repo/issue/choose.tmpl]: Shouldn't that be a part of `options/locale/locale_en-US.ini`?

File: modules/structs/issue.go
Comment (delvh) [inline modules/structs/issue.go]: What I've noticed from having to work with the GitHub templates already:
Collaborators who already have a lot of in-depth knowledge about a project tend to specify their issues with enough detail that they don't necessarily need to input all troubleshooting information.
Having to fill out the template can be pretty tiresome, especially for large projects like Gitea itself, and often leads to entering unnecessary information simply to satisfy the validation.
This means that later on, another value `for-contributors`, for example, could be added.
So, we could think about converting the type to (bool/)`string`, to support more use cases later on.
You don't need to implement the additional functionality I mentioned yet as that can be postponed.
However, I think converting it into a string can already be done in this PR.
As far as I know, it should not make a difference for the yaml parser.

Comment (JakobDev) [reply modules/structs/issue.go]: I agree that having such a setting would be useful, but @lunny requested compatibility with GitHub above, that's the reason it also searches in .github for the config. On GitHub, `blank_issues_enabled` is a boolean. That doesn't mean we can't add our own settings, but if we want compatibility with GitHub, we should not use a different type for the same setting as GitHub. Another setting would be the better solution here.

File: templates/repo/issue/choose.tmpl
Comment (JakobDev) [reply templates/repo/issue/choose.tmpl]: Sorry. I forgot to change it from my tests. I will do it later, when I have time.

File: modules/structs/issue.go
Comment (delvh) [reply modules/structs/issue.go]: But it would accept the same things as GitHub: `true, yes, on, off, no, false`  (-> `strconv.ParseBool`),
and then more (i.e. `for-contributors`).
So it would still be compatible with GitHub?
As far as I know, you don't need `"` around a YAML string…

Comment (JakobDev) [reply modules/structs/issue.go]: >  As far as I know, you don't need `"` around a YAML string…

That's true in most cases, but reserved keywords like `true` or `false` or numbers needs quotes to be accepted as string.

Comment (delvh) [reply modules/structs/issue.go]: Oh. Didn't know that.
But can our YAML Parser perhaps reconvert a boolean to a string?

Comment (JakobDev) [reply modules/structs/issue.go]: I don't know if the parser could do that. It looks like it needs to know the types of the values. But even when, it doesn't change the fact, that the result would not be compatible with GitHub, which don't accept something like `for-contributors`. It's way better and easier to just add another setting like `enable_blank_issues_for_contributors`.

Comment (jolheiser) [reply modules/structs/issue.go]: fwiw the parser seems to gracefully handle it, though I would need to test it in this context specifically.
https://go.dev/play/p/jNdWPWtK-Jz

Comment (delvh) [reply modules/structs/issue.go]: > It's way better and easier to just add another setting like enable_blank_issues_for_contributors.

I strongly disagree there.
That simply allows for an ungodly amount of edge cases and illegal states.

Comment (delvh) [reply modules/structs/issue.go]: > fwiw the parser seems to gracefully handle it, though I would need to test it in this context specifically. https://go.dev/play/p/jNdWPWtK-Jz

That's exactly what I hoped for.
This means we can still convert it without repercussions.
But I agree with @jolheiser, we should certainly ensure with a test that this behavior is fulfilled.

Comment (JakobDev) [reply modules/structs/issue.go]: I still think this is a bad idea. Even is that could parsed, this will confuse users and is still not comptaible with GitHub. I will keep the current behaviour for this PR. It is better to do is in another PR.

**Comment by JakobDev** [comment]
I had now added a API endpoint to get the issue config. The error text is now translatable.

File: modules/context/repo.go
Comment (wolfogre) [inline modules/context/repo.go]:
```suggestion
	issueConfig := GetDefaultIssueConfig()

	if r.GitRepo == nil {
		return issueConfig, nil
	}

	treeEntry, err := commit.GetTreeEntryByPath(path)
	if err != nil {
		return issueConfig, err
	}

	reader, err := treeEntry.Blob().DataAsync()
	if err != nil {
		log.Debug("DataAsync: %v", err)
		return issueConfig, nil
	}

	configContent, err := io.ReadAll(reader)
	if err != nil {
		return issueConfig, err
	}

	return issueConfig, yaml.Unmarshal(configContent, &issueConfig)
```

File: modules/context/repo.go
Comment (wolfogre) [inline modules/context/repo.go]:
```suggestion
		if _, err := commit.GetTreeEntryByPath(configName + ".yaml"); err == nil {
			return ctx.Repo.GetIssueConfig(configName + ".yaml", commit)
		}

		if _, err := commit.GetTreeEntryByPath(configName + ".yml"); err == nil {
			return ctx.Repo.GetIssueConfig(configName + ".yml", commit)
		}
```

File: modules/context/repo.go
Comment (JakobDev) [reply modules/context/repo.go]: GetDefaultIssueConfig() is only called when using return, so it is only called once. I don't think this is needed.

File: modules/context/repo.go
Comment (lunny) [inline modules/context/repo.go:1127]: defer reader.Close()

File: modules/structs/issue.go
Comment (lunny) [reply modules/structs/issue.go]: We can just keep it only allow `true` and `false` in this PR. And in future PR, it can be allowed other values but we don't need to discuss it in this PR.

File: routers/api/v1/repo/repo.go
Comment (delvh) [inline routers/api/v1/repo/repo.go]: Erm… Given that you now offer an API route, I don't think we can afford not to convert it to `string`, otherwise we **will** get backwards compatibility problems.
Also, we could think about adding a query param that controls which branch to query.

**Review by delvh — CHANGES_REQUESTED** [review]

File: routers/api/v1/repo/repo.go
Comment (JakobDev) [reply routers/api/v1/repo/repo.go]: I'm still thinking that converting to string is a bad idea. Adding a second boolean setting is much better IMHO.

**Comment by JakobDev** [comment]
I have no implemented contact links. I had planned t do this in a separate PR, but this PR is now open for quite a while, so I think it's better to push the code in this PR instead of keeping it on my PC for months until this PR is finally merged.

@wolfogre I haven't found a Way to make YAML keys required. Did you have a Idea?

**Comment by wolfogre** [comment]
> I haven't found a Way to make YAML keys required. Did you have a Idea?

I think you can:

- add `func (*IssueConfigContactLink) Validate() error` to check if required keys are missing;
- add `func (*IssueConfig) Validate() error` to call `ContactLinks[0..len-1].Validate()`;
- check `issueConfig.Validate()` after `yaml.Unmarshal` at `GetIssueConfig`, and return error if there's one.

File: routers/api/v1/repo/repo.go
Comment (wolfogre) [reply routers/api/v1/repo/repo.go]: > ... we will get backwards compatibility problems.

I don't get it, like what? 

> ... adding a query param that controls which branch to query.

It is not correct, the `IssueConfig` only exists in the default branch, if you add a `config.yaml` in another branch, it's just a file and won't work, only the `config.yaml` in default branch is `IssueConfig`.

File: options/locale/locale_en-US.ini
Comment (wolfogre) [inline options/locale/locale_en-US.ini]:
```suggestion
issues.choose.invalid_config = The issue config contains errors:
```

IMO:

- Not `your`, the issue poster may do not own the repo.
- It's not `issue_config.yaml`.

**Comment by JakobDev** [comment]
Thanks for your answer. I hoped there was a better Way.

File: docs/content/doc/usage/issue-pull-request-templates.en-us.md
Comment (wolfogre) [inline docs/content/doc/usage/issue-pull-request-templates.en-us.md]:
```suggestion
```

**Review by wolfogre — APPROVED** [review]
Maybe there's still some room for improvement, but I don't think it's necessary to keep blocking this PR. It works fine in my local env.

**Comment by JakobDev** [comment]
@lunny Any chance this will be merged before 1.19?

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
Need @delvh approval.

File: modules/context/repo.go
Comment (delvh) [inline modules/context/repo.go:1151]: Wait, `about` is required?
I would have thought it is optional…

File: modules/structs/issue.go
Comment (delvh) [inline modules/structs/issue.go]: `IssueConfigValidation`?

File: modules/structs/issue.go
Comment (delvh) [inline modules/structs/issue.go:193]: Should we perhaps omit the `IssueConfig-` prefix?

**Comment by JakobDev** [comment]
Ping @delvh

**Comment by wxiaoguang** [comment]
Happened to see this PR ..... maybe the maintainer team need some mechanisms to handle the "pending reviews"?

Should I help to call the TOC (again)?

**Comment by lunny** [comment]
Need @delvh 's review.

**Review by delvh — COMMENTED** [review]
Yep, sorry for the late reply, my TODO list is way too long already.
Furthermore, I'm still not quite a fan of using a `boolean` where I think a string is more fitting.
However, I can at least accept this PR assuming everyone knows that this is probably not the best option for a long-term solution.

File: modules/context/repo.go
Comment (JakobDev) [reply modules/context/repo.go:1151]: Yes, it is required. If you don't have it on GitHub, the Link will just be ignored.

File: modules/structs/issue.go
Comment (JakobDev) [reply modules/structs/issue.go:193]: This struct is only used as Part of the Issue Config, so I think it makes sense to keep the Prefix to indicate that

File: modules/structs/issue.go
Comment (JakobDev) [reply modules/structs/issue.go]: Done

**Review by delvh — APPROVED** [review]
I've now tested the functionality.
A few closing remarks:
- I think the API route can be either under `repo` or under `issue`. I can certainly see why it's currently under repo. GitHub doesn't offer this API at the moment, so I'm fine with both
- I am not a fan of having `_` in URLs but seeing that other URLs also have the `_` already, I'll allow it
- There should be a divider between the issue templates and the contact links, however, this can be postponed for a later PR:
image


I'm still not a fan of using a `boolean` for that struct, but other than that I don't see anything holding this PR back.

**Comment by JakobDev** [comment]
> There should be a divider between the issue templates and the contact links, however, this can be postponed for a later PR:

I personally prefer it when it has no divider and everything looks the same (except for the buttons) in the box.

> I'm still not a fan of using a `boolean` for that struct, but other than that I don't see anything holding this PR back.

I still think, it's better to add a extra setting. A string that is a boolean is confusing and can cause problems with the API.

**Comment by jolheiser** [comment]
🎺 🤖

**Pull request merged**
